packages/peregrine/src/RestApi/Magento2/M2ApiRequest.js

```diff
@@ -0,0 +1,264 @@
+        this.resourceUrl = resourceUrl;
+        // merge headers specially
+        this.opts = {
+            // can be overridden
+            method: 'GET',
+            signal: this.controller.signal,
+            credentials: 'include',
+            ...opts,
+            // cannot be overridden, only appended to
+            headers: withDefaultHeaders(opts.headers)
+        };
+    }
+    /**
+     * Execute the request. Must be run before {@link M2ApiRequest#getResponse}
+     * or {@link M2ApiRequest#cancel} can be called.
+     */
+    run() {
+        if (this._isMulticastable()) {
+            this._promise = this._fetchMulticast();
+        } else {
+            this._promise = this._fetch();
+        }
+    }
+    /**
+     * Get the promise for the network operation. Can only be called after
+     * `.run()` is called.
+     * For multicast requests, will return a promise for a new copy of the
+     * response every time it is called, since a Body can only be used once.
+     * Exists so that requests can reuse the promises from other requests.
+     * @returns {Promise} Promise for the result of the request.
+     */
+    getResponse() {
+        if (!this._promise) {
+            throw new Error(
+                'M2ApiRequest#getResponse() called before M2ApiRequest#run(), so no promise exists yet'
+            );
+        }
+        if (this._isMulticastable()) {
+            return this._promise.then(res => res.clone());
+        } else {
+            return this._promise;
+        }
+    }
+    /**
+     * Abort the network operation. Multicasted requests catch the AbortError
+     * and attempt to reuse a more recent matching request from cache. Other
+     * requests will pass the AbortError rejection through to the consumer.
+     */
+    abortRequest() {
+        this.controller.abort();
+    }
+    /**
+     * Check if this request intends to override prior requests to the same
+     * resource. Rolling requests will take the place of prior outstanding
+     * requests, to ensure the freshest resource at the cost of additional
+     * network calls.
+     *
+     * The current logic for rolling requests is determined by the `cache`
+     * option. [Cache modes](https://developer.mozilla.org/en-US/docs/Web/API/Request/cache)
+     * `reload` and `no-store` both indicate complete cache bypass. This
+     * logically implies that the user has just changed server state and wants
+     * to force retrieve an updated resource, so multicasting a prior request
+     * would not be appropriate--the response may not reflect the more recent
+     * change.
+     * @returns {boolean} True if the request is rolling.
+     */
+    isRolling() {
+        return this.opts.cache === 'no-store' || this.opts.cache === 'reload';
+    }
+    /**
+     * Make sure not to multicast POST requests which have a nonempty body,
+     * since they are unsafe and non-idempotent, so each call may mutate
+     * server-side state.
```

Comment:
Is this something that can be caught as an error or a warning, such that it might appear in a console? If so, is that happening yet?

Reply:
In this comment I mean [the REST definition of the term] 'unsafe' (http://restcookbook.com/HTTP%20Methods/idempotency/), which is not a bad condition; it just shouldn't be multicast. It can be overridden, but I don't think a warning is necessary. I figure if a dev forces `multicast: true` on a request, they know what they're doing.